workspace_file_harvester/app.py

```
@@ -0,0 +1,168 @@
+        latest_harvested = {}
+
+        logging.error(f"Harvesting from {workspace_name} {source_s3_bucket}")
+
+        pulsar_client = get_pulsar_client()
+        producer = pulsar_client.create_producer(
+            topic="harvested",
+            producer_name=f"workspace_file_harvester/{workspace_name}_{uuid.uuid1().hex}",
+            chunking_enabled=True,
+        )
+
+        file_harvester_messager = FileHarvesterMessager(
+            s3_client=s3_client,
+            output_bucket=target_s3_bucket,
+            cat_output_prefix=f"git-harvester/user-datasets/{workspace_name}/",
+            producer=producer,
+            workspace_name=workspace_name,
+        )
+
+        metadata_s3_key = f"harvested-metadata/file-harvester/{workspace_name}"
+        previously_harvested, last_modified = get_metadata(
+            target_s3_bucket, metadata_s3_key, s3_client
+        )
+        file_age = datetime.datetime.now() - last_modified
+        if file_age < datetime.timedelta(
+            seconds=int(os.environ.get("RUNTIME_FREQUENCY_LIMIT", "6"))
+        ):
+            logging.error(f"Harvest not completed - previous harvest was {file_age} seconds ago")
+            return JSONResponse(
+                content={"message": "Wait a while before trying again"}, status_code=429
+            )
+        logging.error(f"Previously harvested URLs: {previously_harvested}")
+
+        count = 0
+        for details in s3_client.list_objects(
+            Bucket=source_s3_bucket, Prefix=f"{workspace_name}/eodh-config/"
+        ).get("Contents", []):
+            key = details["Key"]
+            if not key.endswith("/"):
+                logging.error(f"{key} found")
+                data, _ = get_file_s3(source_s3_bucket, key, s3_client)
+
+                previous_hash = previously_harvested.pop(key, None)
+
+                file_hash = get_file_hash(json.dumps(data))
+                if not previous_hash or previous_hash != file_hash:
```

Instead of doing this you could record the ETag from the previous call, then use `get_object(Bucket=..., Key=..., IfNoneMatch=previous_etag)` (with luck the default value for IfNoneMatch is None to make it even easier). If `get_object` returns `304 Not Modified` then the file hasn't changed and it doesn't get downloaded.

Ignoring the error handling and functions, you're then doing something similar to this

```
previous_etag = previously_harvested.pop(key, None)
file_obj = s3_client.get_object(Bucket=bucket, Key=key, IfNoneMatch=previous_etag)
if file_obj["ResponseMetadata"]["HTTPStatusCode"] != 304:
    harvested_data[key] = file_obj["Body"].read().decode("utf-8")
    latest_harvested[key] = file_obj["ETag"]
else:
    latest_harvested[key] = previous_etag
```

---

Added